include/deal.II/meshworker/scratch_data.h

```diff
@@ -0,0 +1,1192 @@
+     */
+    const std::vector<types::global_dof_index> &
+    get_neighbor_dof_indices() const;
+
+    /**
+     * Return a reference to the used mapping.
+     */
+    const Mapping<dim, spacedim> &
+    get_mapping() const;
+
+  private:
+    /**
+     * The mapping used by the internal FEValues. Make sure it lives
+     * longer than this class.
+     */
+    SmartPointer<const Mapping<dim, spacedim>> mapping;
+
+    /**
+     * The finite element used by the internal FEValues. Make sure it lives
+     * longer than this class.
+     */
+    SmartPointer<const FiniteElement<dim, spacedim>> fe;
+
+    /**
+     * Quadrature formula used to integrate on the current cell, and on its
+     * neighbor.
+     */
+    Quadrature<dim> cell_quad;
+
+    /**
+     * Quadrature formula used to integrate on faces, subfaces, and neighbor
+     * faces and subfaces.
+     */
+    Quadrature<dim - 1> face_quad;
+
+    /**
+     * UpdateFlags to use when initializing the cell FEValues object.
+     */
+    UpdateFlags cell_flags;
+
+    /**
+     * UpdateFlags to use when initializing the neighbor cell FEValues objects.
+     */
+    UpdateFlags neighbor_cell_flags;
+
+    /**
+     * UpdateFlags to use when initializing FEFaceValues and FESubfaceValues
+     * objects.
+     */
+    UpdateFlags face_flags;
+
+    /**
+     * UpdateFlags to use when initializing neighbor FEFaceValues and
+     * FESubfaceValues objects.
+     */
+    UpdateFlags neighbor_face_flags;
+
+    /**
+     * Finite element values on the current cell.
+     */
+    std::unique_ptr<FEValues<dim, spacedim>> fe_values;
+
+    /**
+     * Finite element values on the current face.
+     */
+    std::unique_ptr<FEFaceValues<dim, spacedim>> fe_face_values;
+
+    /**
+     * Finite element values on the current subface.
+     */
+    std::unique_ptr<FESubfaceValues<dim, spacedim>> fe_subface_values;
+
+    /**
+     * Finite element values on the neighbor cell.
+     */
+    std::unique_ptr<FEValues<dim, spacedim>> neighbor_fe_values;
+
+    /**
+     * Finite element values on the neighbor face.
+     */
+    std::unique_ptr<FEFaceValues<dim, spacedim>> neighbor_fe_face_values;
+
+    /**
+     * Finite element values on the neighbor subface.
+     */
+    std::unique_ptr<FESubfaceValues<dim, spacedim>> neighbor_fe_subface_values;
+
+    /**
+     * Dof indices on the current cell.
+     */
+    std::vector<types::global_dof_index> local_dof_indices;
+
+    /**
+     * Dof indices on the neighbor cell.
+     */
+    std::vector<types::global_dof_index> neighbor_dof_indices;
+
+    /**
+     * Arbitrary user data, identified by a string.
+     */
+    std::map<std::string, boost::any> any_data;
+  };
+
+
+
+  //
+  // Inline and template methods
+  //
+  template <int dim, int spacedim>
+  template <typename type>
+  void
+  ScratchData<dim, spacedim>::add_copy(const type &       entry,
+                                       const std::string &name)
+  {
+    any_data[name] = entry;
```

Comment:
Why would you want to disallow overwriting?

Reply:
Since this is a public method, this is the situation I would want to help avoid:
```
ScratchData scratch (...);
scratch.add_copy (solution_t0,"solution_t0");
scratch.add_copy (solution_t1,"solution_t0"); // Whoops, typo in name!
MeshWorker::mesh_loop(...,scratch,...);
```
Would you that it might be appropriate to have two variants of these functions which could help avoid this scenario. Something like `add_unique_copy()`, `add_or_overwrite_copy()` (this being the current function), `add_unique_reference()`, `add_or_overwrite_reference()`. To me this is like "const correctness" for the imported data.

Reply:
@luca-heltai Any comment on this?